### ISSUE TITLE
Investigate reducing cost of initializing editor contributions 

### ISSUE THREAD

**Issue by mjbvz** [microsoft/vscode#164171]
On startup with an untitled editor visible, we create around 32 editor contributions. This initializing has an upfront cost, and also means that we end up registering more event subscribes that need to be notified in the future. Some (or most) of this initialization work can be deferred until the controller's functionality is actually needed

Here's a performance profile editor contribution initialization. In this case it took around 20ms total on a fast machine, which isn't a huge number but it delays the time until the editor becomes visible.

- [ ] 6.25% `SuggestController` — `suggestController.ts:121` @jrieken 
- [ ] 4.69% `StickyScrollControllers` — `tickyScrollController.ts:28` @jrieken 
- [ ] 3.91% `TestingDecorations` — `testingDecorations.ts:329` @connor4312 
- [ ] 3.91% `TestingOutputPeekController` — `testingOutputPeek.ts:426` @connor4312 
- [ ] 3.13% `FindController` — `findController.ts:422`
- [ ] 3.13% `ColorDetector` — `colorDetector.ts:50`
- [ ] 3.13% `LinkDetector` — `links.ts:52`
- [x] 3.13% `ParameterHintsController` — `parameterHints.ts:32` @mjbvz  https://github.com/microsoft/vscode/pull/164112
- [ ] 2.34% `DirtyDiffController` — `dirtydiffDecorator.ts:579`
- [ ] 2.34% `CodeLensContribution` — `codelensController.ts:46`
- [ ] 2.34% `LinkedEditingContribution` — `linkedEditing.ts:82`
- [ ] 2.34% `InlayHintsController` — `inlayHintsController.ts:116`
- [ ] 2.34% `CommentController` — `commentsEditorC…ibution.ts:339`
- [ ] 1.56% `FoldingController` — `folding.ts:111`
- [ ] 1.56% `AcceptChangesContribution` — `userDataSyncCon…ctsView.ts:238`
- [ ] 1.56% `UnicodeHighlighter` — `unicodeHighlighter.ts:47`
- [x] 1.56% `CodeActionController` — `codeActionCommands.ts:98`  @mjbvz 
- [x] 1.56% `AccessibilityHelpController` — `accessiblity.ts` @mjbvz https://github.com/microsoft/vscode/pull/164166/files

Lower impact: 

- [ ] 0.78% `CallHierarchyController` — `callHierarchy.c…ribution.ts:58`
- [ ] 0.78% `DefineKeybindingController` — `keybindingsEdit…ribution.ts:51`
- [ ] 0.78% `SelectionHighlighter` — `multicursor.ts:853`
- [ ] 0.78% `TabCompletionController` — `tabCompletion.ts:45`
- [x] 0.78% `DropIntoEditorController` — `dropIntoEditorC…ribution.ts:34`
- [ ] 0.78% `UnusualLineTerminatorsDetector` — `unusualLineTerminators.ts:33`
- [ ] 0.78% `GhostTextController` — `ghostTextController.ts:42`
- [ ] 0.78% `LazyDebugEditorContribution` — `debugEditorContribution.ts:215`
- [ ] 0.78% `LazyBreakpointEditorContribution` — `breakpointEdito…ibution.ts:194`
- [ ] 0.78% `ModesHoverController` — `hover.ts:51`
- [ ] 0.78% `GotoDefinitionAtPositionEditorContribution` — `goToDefinitionAtPosition.ts:49`
- [ ] 0.78% `FloatingClickMenu` — `codeeditor.ts:220`
- [ ] 0.78% `SettingsEditorContribution` — `preferencesEditor.ts:21`
- [ ] 0.78% `EditorContribution` — `suggestInlineCo…letions.ts:234`
- [ ] 0.78% `CopyPasteController` — `copyPasteController.ts:53`
- [ ] 0.78% `CursorUndoRedoController` — `cursorUndoRedo.ts:60`
- [ ] 0.78% `CursorUndoRedoController` — `cursorUndoRedo.ts:60`

---

As a first step, please review your contribution and try to defer any work that is not needed immediately. The creation of UI is a good candidate for this, as often the UI is only shown later when a specific command is triggered or when a language provider updates

In the future, we may also explore being more aggressive about deferring work. Parameter hints functionality is not required instantly on startup, so the entire parameter hints contribution could likely be loaded when editor is idle

**Comment by rebornix** [comment]
@mjbvz when testing in notebook editor, I found after the whole workbench initializes, creating a monaco instance will create 60ish contribs. Thus when creating multiple full feature monaco editors, we kept creating all these contribs in every editor instance and the `ctor` time is pretty significant.

Not all contributions will work in all languages/scenarios, for example, the settings contrib only work in the `settings.json`. I wonder if we could push this a bit further, in addition to asking each contrib to do lazy initialization, we can consider having conditional contrib and load them when needed.
